bindings/python/multibody/joint/joints-models.hpp

```diff
@@ -54,41 +54,87 @@ namespace pinocchio
 
     // specialization for JointModelComposite
 
-    struct JointModelCompositeAddJointVisitor : public boost::static_visitor<void>
+    struct JointModelCompositeAddJointVisitor : public boost::static_visitor<JointModelComposite &>
     {
       JointModelComposite & m_joint_composite;
       const SE3 & m_joint_placement;
 
       JointModelCompositeAddJointVisitor(JointModelComposite & joint_composite,
-                      const SE3 & joint_placement)
+                                         const SE3 & joint_placement)
       : m_joint_composite(joint_composite)
       , m_joint_placement(joint_placement)
       {}
 
       template <typename JointModelDerived>
-      void operator()(JointModelDerived & jmodel) const
+      JointModelComposite & operator()(JointModelDerived & jmodel) const
       {
         return m_joint_composite.addJoint(jmodel,m_joint_placement);
       }
     }; // struct JointModelCompositeAddJointVisitor
 
-    static void addJoint_proxy(JointModelComposite & joint_composite,
-                         bp::object jmodel,
-                         const SE3 & joint_placement = SE3::Identity())
+    static JointModelComposite & addJoint_proxy(JointModelComposite & joint_composite,
+                                                const JointModelVariant & jmodel_variant,
+                                                const SE3 & joint_placement = SE3::Identity())
     {
-      JointModelVariant jmodel_variant = bp::extract<JointModelVariant> (jmodel)();
-      boost::apply_visitor(JointModelCompositeAddJointVisitor(joint_composite,joint_placement), jmodel_variant);
+      return boost::apply_visitor(JointModelCompositeAddJointVisitor(joint_composite,joint_placement), jmodel_variant);
     }
 
     BOOST_PYTHON_FUNCTION_OVERLOADS(addJoint_proxy_overloads,addJoint_proxy,2,3)
 
+    struct JointModelCompositeConstructorVisitor : public boost::static_visitor<boost::shared_ptr<JointModelComposite> >
+    {
+      const SE3 & m_joint_placement;
+
+      JointModelCompositeConstructorVisitor(const SE3 & joint_placement)
+      : m_joint_placement(joint_placement)
+      {}
+
+      template <typename JointModelDerived>
+      boost::shared_ptr<JointModelComposite> operator()(JointModelDerived & jmodel) const
```

Comment:
Why shared_ptr?
We must avoid them because they are inefficient, and certainly not useful here.

Reply:
I had problems exposing this constructor.
The C++ signature is the following
```
template<typename JointModel>
JointModelCompositeTpl(const JointModelBase<JointModel> & jmodel,
                                        const SE3 & placement = SE3::Identity())
```
The problem is that to work in python it should accept a `JointModelVariant`, therefore I need a proxy, and this is the way I found to use a proxy for the constructor.
I basically followed the example here
[](https://github.com/boostorg/python/blob/master/test/injected.cpp)
Do you know of any alternatives? I don't want to use a raw pointer, I fear it might cause leaks and all sort of problems

Reply:
The pointer is the right solution

Reply:
I changed to the pointer